### PR TITLE
Implemented order available payment methods and order update payment

### PR DESCRIPTION
**Pull request by dlobato**
Hi,

We have implemented two new actions to handle payment state on placed orders:

1. Get available payment methods: Same as /{channelCode}/checkout/{token}/payment but on order endpoint.
2. Update payment method on unpaid placed orders to implement the same behavior as in ShopBundle to allow a customer to change the payment method while order is unpaid.

1 is just reusing the same action from checkout, not sure if I should reimplement this as a different action.
2 adds validation as I've seen on other controllers, is this the preferred approach?

Tests are failing, there's something weird I couldn't figure out on loading the fixtures. I bet I'm doing something wrong. Am I missing something obvious there?

I've also added a new field to placedorder view: paymentState.

David.

### CODE REVIEW

File: src/Validator/Constraints/OrderExists.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    /** @var array */
+    public $state = [OrderInterface::STATE_NEW, OrderInterface::STATE_FULFILLED, OrderInterface::STATE_CANCELLED];
+
+    public function __construct($options = null)
+    {
+        //translate constant name to actual value
+        if (isset($options['state'])) {
+            $options['state'] = array_map(
+                static function ($i) {
+                    return defined($i) ? constant($i) : $i;
```
Comment (mamazu) [inline src/Validator/Constraints/OrderExists.php]: There is just a limited amour of possibilities, you could do a switch statement here.

File: tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        $placedOrder = $fixtures['placed_order'];
+
+        $data =
+<<<EOT
+        {
+            "method": "PBC"
+        }
+EOT;
+
+        $this->client->request('PUT', $this->getPaymentUrl($placedOrder->getTokenValue()) . '/0', [], [], self::CONTENT_TYPE_HEADER, $data);
```
Comment (mamazu) [inline tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php]: Just use the value that is in the fixture. No need to reference the object.

File: src/Validator/Constraints/OrderExists.php (hunk 1 above)
Comment (dlobato) [reply src/Validator/Constraints/OrderExists.php]: If you think the constant translation is overkill, I can remove it and the code inside the if statement would look like:
`$options['state'] = (array)$options['state']`

Not sure why would we need and switch here, do you mean to validate thet state is a valid order state?

File: tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php (hunk 2 above)
Comment (dlobato) [reply tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php]: Sure, makes sense. Anyway, tests fails because:
ORDERTOKENPLACED has state cart, which hits on the validation rule. But if I change it to new, then other tests (order related) fail because totals don't match anymore... Maybe I should create another test order.

But the second test (ORDERTOKENPAID) throws an exception while loading fixtures, and I can't figure out why, have you ever seen this? I'm doing something wrong, sure, but where?

File: src/Command/Order/UpdatePaymentMethod.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace Sylius\ShopApiPlugin\Command\Order;
+
+class UpdatePaymentMethod
+{
+    /** @var string */
+    protected $orderToken;
+
+    /** @var mixed */
+    protected $paymentIdentifier;
```
Comment (GSadee) [inline src/Command/Order/UpdatePaymentMethod.php]: `$paymentId`?

File: src/Command/Order/UpdatePaymentMethod.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    /** @var string */
+    protected $paymentMethod;
```
Comment (GSadee) [inline src/Command/Order/UpdatePaymentMethod.php]: `$paymentMethodCode` to be more clear?

File: src/Handler/Order/UpdatePaymentMethodHandler.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        $this->paymentMethodRepository = $paymentMethodRepository;
+    }
+
+    public function __invoke(UpdatePaymentMethod $choosePaymentMethod): void
+    {
+        /** @var OrderInterface $order */
+        $order = $this->orderRepository->findOneBy(['tokenValue' => $choosePaymentMethod->orderToken()]);
+
+        Assert::notNull($order, 'Order has not been found.');
+        Assert::eq(OrderPaymentStates::STATE_AWAITING_PAYMENT, $order->getPaymentState(), 'Only awaiting payment orders can be updated.');
```
Comment (GSadee) [inline src/Handler/Order/UpdatePaymentMethodHandler.php]: `eq` => `same`

File: src/Resources/config/routing/order.yml [hunk 6]
```diff
@@ -9,3 +9,15 @@ sylius_shop_api_order_details:
...
+    path: /orders/{token}/payment
+    methods: [GET]
+    defaults:
+        _controller: sylius.shop_api_plugin.controller.checkout.show_available_payment_methods_action
+
+sylius_shop_api_order_update_payment_method:
+    path: /orders/{token}/payment/{paymentId}
+    methods: [PUT]
+    defaults:
+        _controller: sylius.shop_api_plugin.controller.order.update_payment_method_action
```
Comment (GSadee) [inline src/Resources/config/routing/order.yml]: Missing blank line at the end of file

File: src/Validator/Constraints/OrderExists.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+                },
+                (array) $options['state']
+            );
+        }
+        parent::__construct($options);
+    }
+
+    /** {@inheritdoc} */
+    public function getTargets()
```
Comment (GSadee) [inline src/Validator/Constraints/OrderExists.php]: You can add `string` return type and docblock is not necessary IMO

File: tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    /**
+     * @test
+     */
+    public function it_does_not_allow_to_update_payment_method_on_paid_order(): void
+    {
+        $fixtures = $this->loadFixturesFromFiles(['customer.yml', 'country.yml', 'address.yml', 'shop.yml', 'payment.yml', 'shipping.yml', 'order.yml']);
+        $this->logInUser('hater@queen.com', '123password');
+
+        $data =
+            <<<EOT
```
Comment (GSadee) [inline tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php]: Wrong indentation

File: tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        $response = $this->client->getResponse();
+        $this->assertResponseCode($response, Response::HTTP_BAD_REQUEST);
+    }
+
+    /**
+     * @param string $token
+     *
+     * @return string
+     */
+    private function getPaymentUrl($token): string
```
Comment (GSadee) [inline tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php]: `private function getPaymentUrl(string $token): string` and docblock is not needed

File: src/Command/Order/UpdatePaymentMethod.php (hunk 3 above)
Comment (dlobato) [reply src/Command/Order/UpdatePaymentMethod.php]: copy paste from here \Sylius\ShopApiPlugin\Command\Cart\ChoosePaymentMethod::$paymentIdentifier

File: src/Command/Order/UpdatePaymentMethod.php (hunk 4 above)
Comment (dlobato) [reply src/Command/Order/UpdatePaymentMethod.php]: same as above: \Sylius\ShopApiPlugin\Command\Cart\ChoosePaymentMethod::$paymentMethod
but sure I can update those

**Comment by dlobato** [comment]
@GSadee @mamazu I'm struggling here with the tests... Can't figure out what's wrong. If I run the tests alone on local they work fine, but when running them one after the other there's an unrelated error. Seems to me some sort of race condition while loading the fixtures for the second time. Is something obvious that I'm not seeing here?

File: src/Validator/Constraints/OrderExists.php (hunk 1 above)
Comment (mamazu) [reply src/Validator/Constraints/OrderExists.php]: What should this do? The job of the constraint is not to translate the state it is given.

Comment (dlobato) [reply src/Validator/Constraints/OrderExists.php]: fair point, I'll simplify this

File: tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php (hunk 2 above)
Comment (mamazu) [reply tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php]: Yeah creating a new order for that matter would be a good idea to not interfere with other tests.

About the fixture loading, I haven't seen this issue. But I can try to look into it.

File: src/Validator/Constraints/OrderExists.php [hunk 10]
```diff
@@ -13,28 +13,9 @@ final class OrderExists extends Constraint
     /** @var string */
     public $message = 'sylius.shop_api.order.not_exists';
 
-    /** @var array */
+    /** @var mixed */
```
Comment (mamazu) [inline src/Validator/Constraints/OrderExists.php]: Why did you change that? Isn't the value an array?

Comment (dlobato) [reply src/Validator/Constraints/OrderExists.php]: No, it can be just a string or an array of strings. The constraint mapping xml accepts this two ways of passing options:
```
<option name="state">new</option>
```
or
```
<option name="state">
   <value>new</value>
   <value>other</value>
</option>
```

Comment (mamazu) [reply src/Validator/Constraints/OrderExists.php]: Yeah, I already feared so then I would suggest: `/** @var string|array<string> */` so that we have at least some kind of idea what we are dealing with (and it helps the static analysis).

File: tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php (hunk 2 above)
Comment (dlobato) [reply tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php]: There two new orders with required information.

Comment (dlobato) [reply tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php]: https://github.com/Sylius/ShopApiPlugin/pull/422/files#diff-2170a1f75ea7146d4434637215da40b7

**Comment by dlobato** [comment]
I fixed the tests by changing the item used on the order. Sadly, I couldn't figure out what was the problem but it is related to mug item having product attributes. Since the items on the order are irrelevant for the test purposes I've just used another one and updated the expected responses accordingly.

@mamazu @GSadee is there anything else to code review?

**Comment by mamazu** [comment]
Ill have a look at it this evening.

File: spec/Handler/Order/UpdatePaymentMethodHandlerSpec.php [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+class UpdatePaymentMethodHandlerSpec extends ObjectBehavior
+{
+    function let(
+        OrderRepositoryInterface $orderRepository,
+        PaymentMethodRepositoryInterface $paymentMethodRepository
+    ): void {
+        $this->beConstructedWith($orderRepository, $paymentMethodRepository);
+    }
+
+    function it_assigns_choosen_payment_method_to_specified_payment(
```
Comment (mamazu) [inline spec/Handler/Order/UpdatePaymentMethodHandlerSpec.php]: choosen => chosen

File: src/Handler/Order/UpdatePaymentMethodHandler.php [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        PaymentMethodRepositoryInterface $paymentMethodRepository
+    ) {
+        $this->orderRepository = $orderRepository;
+        $this->paymentMethodRepository = $paymentMethodRepository;
+    }
+
+    public function __invoke(UpdatePaymentMethod $choosePaymentMethod): void
+    {
+        /** @var OrderInterface $order */
+        $order = $this->orderRepository->findOneBy(['tokenValue' => $choosePaymentMethod->orderToken()]);
```
Comment (mamazu) [inline src/Handler/Order/UpdatePaymentMethodHandler.php]: This call will also return carts. You need to check the state as well.

**Comment by mamazu** [comment]
There are still some open conversations if you could resolve that, this would be great.

File: src/Handler/Order/UpdatePaymentMethodHandler.php (hunk 12 above)
Comment (dlobato) [reply src/Handler/Order/UpdatePaymentMethodHandler.php]: Now I check order state is not cart. I've updated phpspec tests accordingly.

File: doc/swagger.yml [hunk 13]
```diff
@@ -1009,6 +1009,53 @@ paths:
                     description: "Order with given tokenValue not found"
             security:
                 -   bearerAuth: []
+    /{channelCode}/orders/{token}/payment:
```
Comment (lchrusciel) [inline doc/swagger.yml]: `payment` -> `payments` ?

File: spec/Command/Order/UpdatePaymentMethodSpec.php [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    function it_throws_an_exception_if_order_token_is_not_a_string(): void
+    {
+        $this->beConstructedWith(new \stdClass(), 1, 'COD_METHOD');
+
+        $this->shouldThrow(TypeError::class)->duringInstantiation();
+    }
+
+    function it_throws_an_exception_if_payment_method_code_is_not_a_string(): void
```
Comment (lchrusciel) [inline spec/Command/Order/UpdatePaymentMethodSpec.php]: Since the introduction of strict type, we can remove TypeError specs

File: spec/Handler/Order/UpdatePaymentMethodHandlerSpec.php [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+use Prophecy\Argument;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+use Sylius\Component\Core\OrderPaymentStates;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Sylius\Component\Core\Repository\PaymentMethodRepositoryInterface;
+use Sylius\ShopApiPlugin\Command\Order\UpdatePaymentMethod;
+
+class UpdatePaymentMethodHandlerSpec extends ObjectBehavior
```
Comment (lchrusciel) [inline spec/Handler/Order/UpdatePaymentMethodHandlerSpec.php]: specs should be final

File: tests/DataFixtures/ORM/order.yml [hunk 16]
```diff
@@ -49,6 +70,11 @@ Sylius\Component\Core\Model\Payment:
         method: "@pay_by_check"
         order: "@placed_order"
         currencyCode: "GBP"
+    another_placed_order_payment:
+        state: "cart"
```
Comment (lchrusciel) [inline tests/DataFixtures/ORM/order.yml]: `placed_order_payment` should be in `new` state

File: src/Handler/Order/UpdatePaymentMethodHandler.php [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+        /** @var PaymentMethodInterface $paymentMethod */
+        $paymentMethod = $this->paymentMethodRepository->findOneBy(['code' => $choosePaymentMethod->paymentMethod()]);
+
+        Assert::notNull($paymentMethod, 'Payment method has not been found');
+        Assert::true(isset($order->getPayments()[$choosePaymentMethod->paymentIdentifier()]), 'Can not find payment with given identifier.');
+
+        $payment = $order->getPayments()[$choosePaymentMethod->paymentIdentifier()];
+
+        $payment->setMethod($paymentMethod);
```
Comment (lchrusciel) [inline src/Handler/Order/UpdatePaymentMethodHandler.php]: We should also assert that payment is in `new` state

File: doc/swagger.yml (hunk 13 above)
Comment (dlobato) [reply doc/swagger.yml]: I'd rather keep it like this to be homogeneous with /{channelCode}/checkout/{token}/payment

File: tests/DataFixtures/ORM/order.yml (hunk 16 above)
Comment (dlobato) [reply tests/DataFixtures/ORM/order.yml]: I've just updated another_placed_order_payment as placed_order_payment was already there.

File: spec/Validator/Order/OrderExistsValidatorSpec.php [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+            ])->willReturn($order);
+
+        $executionContext->addViolation(Argument::any())->shouldNotBeCalled();
+
+        $this->validate('ORDERTOKEN', new OrderExists(
+            ['state' => [OrderInterface::STATE_NEW, 'other_state']]
+        ));
+    }
+
+    function it_adds_constraint_if_order_does_not_exits_exists(
```
Comment (GSadee) [inline spec/Validator/Order/OrderExistsValidatorSpec.php]: `it_adds_constraint_if_order_does_not_exist`

File: src/Command/Order/UpdatePaymentMethod.php (hunk 4 above)
Comment (GSadee) [reply src/Command/Order/UpdatePaymentMethod.php]: I would be for changing that in the added classes in this PR (and open a new PR for the changes in existing classes)

**Comment by lchrusciel** [comment]
Hello David,

Sorry for such a long response time. I found some time to dig into ShopAPI and recently fix the way, how we handle placed orders in the test environment (https://github.com/Sylius/ShopApiPlugin/pull/439/commits/0e445e4b876e96139e45a53a5e98c8c87fa745f5)

This made this PR conflicted. Would you like to fix it?

**Comment by dlobato** [comment]
@lchrusciel I've updated the tests. But now I can't set the payment state to something different from awaiting_payment to test it_does_not_allow_to_update_payment_method_on_paid_order, any suggestion?
I'd say that I need to trigger pay, but not sure how...

**Comment by mamazu** [comment]
You can not set the payment method once the payment is completed.

**Comment by mamazu** [comment]
Ah, I see your problem. Currently there is no way for the API to mark an order as payed. What you can do is try to generate a fixture with a payed order. Other than that I don't know see #310

**Comment by jdeveloper** [comment]
What's the status of this issue?

Is there any hope to get merged? It would be great. So if I choose a payment method, paypal for example, complete the order but then I cancel in paypal, I could change the payment method.

**Comment by mamazu** [comment]
I have rebased the issue but I can not push it. @dlobato Please allow edits from maintainers to your branch.

**Comment by dlobato** [comment]
@mamazu I don't know how to enable that. I was trying this: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork But I don't see said checkbox anywhere...

**Comment by mamazu** [comment]
Hm, very strange. I can push to other merge request. Anyhow I have pushed my version of the branch to my repository [mamazu/handle_order_payment](https://github.com/mamazu/ShopApiPlugin/tree/handle_order_payment). You can run the following commands to get this merge request up to speed:
```bash
# Add me as a remote
git remote add mamazu git@github.com:mamazu/ShopApiPlugin.git

# Fetch all the branches
git fetch mamazu

# If you are not on your local branch change to it. Otherwise just run
git reset --hard mamazu/handle_order_payment

# Push the new changes to git (you might need to explicitly add `origin handle_order_payment`)
git push --force-with-lease
```

**Comment by dlobato** [comment]
Your changes are already on my branch...

**Comment by mamazu** [comment]
This can't be I have fixed the tests (eg. removed the channelcode from the route).

**Comment by dlobato** [comment]
Check it out now

**Comment by mamazu** [comment]
Looks good. Can you run `composer fix` and then commit the changes. Then the merge request ready for merge.

File: src/Request/Order/UpdatePaymentMethodRequest.php [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        return new UpdatePaymentMethod($this->token, $this->paymentIdentifier, $this->paymentMethod);
+    }
+
+    public function getOrderToken(): string
+    {
+        return $this->token;
+    }
+
+    public function getPaymentId()
```
Comment (mamazu) [inline src/Request/Order/UpdatePaymentMethodRequest.php]:
```suggestion
    /** @return int|string */
    public function getPaymentId()
```

File: src/Command/Order/UpdatePaymentMethod.php [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+    /** @var string */
+    protected $orderToken;
+
+    /** @var mixed */
+    protected $paymentId;
+
+    /** @var string */
+    protected $paymentMethodCode;
+
+    public function __construct(string $orderToken, $paymentId, string $paymentMethodCode)
```
Comment (mamazu) [inline src/Command/Order/UpdatePaymentMethod.php]:
```suggestion
    /** @param int|string $paymentId */
    public function __construct(string $orderToken, $paymentId, string $paymentMethodCode)
```

File: src/Command/Order/UpdatePaymentMethod.php [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+        $this->paymentId = $paymentId;
+        $this->paymentMethodCode = $paymentMethodCode;
+    }
+
+    public function orderToken(): string
+    {
+        return $this->orderToken;
+    }
+
+    public function paymentId()
```
Comment (mamazu) [inline src/Command/Order/UpdatePaymentMethod.php]:
```suggestion
    /** @return string|int */
    public function paymentId()
```

File: src/Validator/Order/OrderExistsValidator.php [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+
+    public function __construct(OrderRepositoryInterface $orderRepository)
+    {
+        $this->orderRepository = $orderRepository;
+    }
+
+    public function validate($token, Constraint $constraint): void
+    {
+        if (null === $this->orderRepository->findOneBy(['tokenValue' => $token, 'state' => $constraint->state])) {
+            $this->context->addViolation($constraint->message);
```
Comment (mamazu) [inline src/Validator/Order/OrderExistsValidator.php]:
```suggestion
            /** @var Sylius\ShopApiPlugin\Validator\Constraints\OrderExists $constraint */
            $this->context->addViolation($constraint->message);
```

File: src/Validator/Order/OrderExistsValidator.php [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+
+    public function __construct(OrderRepositoryInterface $orderRepository)
+    {
+        $this->orderRepository = $orderRepository;
+    }
+
+    public function validate($token, Constraint $constraint): void
+    {
+        if (null === $this->orderRepository->findOneBy(['tokenValue' => $token, 'state' => $constraint->state])) {
+            /** @var Sylius\ShopApiPlugin\Validator\Constraints\OrderExists $constraint */
```
Comment (mamazu) [inline src/Validator/Order/OrderExistsValidator.php]:
```suggestion
```

File: src/Validator/Order/OrderExistsValidator.php [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+    private $orderRepository;
+
+    public function __construct(OrderRepositoryInterface $orderRepository)
+    {
+        $this->orderRepository = $orderRepository;
+    }
+
+    public function validate($token, Constraint $constraint): void
+    {
+        if (null === $this->orderRepository->findOneBy(['tokenValue' => $token, 'state' => $constraint->state])) {
```
Comment (mamazu) [inline src/Validator/Order/OrderExistsValidator.php]:
```suggestion
        /** @var Sylius\ShopApiPlugin\Validator\Constraints\OrderExists $constraint */
        if (null === $this->orderRepository->findOneBy(['tokenValue' => $token, 'state' => $constraint->state])) {
```

**Comment by mamazu** [comment]
Can you please checkout the code and put the right doc comment in there.

**Comment by mamazu** [comment]
Thank you, David! :1st_place_medal:

File: src/Handler/Order/UpdatePaymentMethodHandler.php [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+        $this->paymentMethodRepository = $paymentMethodRepository;
+    }
+
+    public function __invoke(UpdatePaymentMethod $choosePaymentMethod): void
+    {
+        /** @var OrderInterface $order */
+        $order = $this->orderRepository->findOneBy(['tokenValue' => $choosePaymentMethod->orderToken()]);
+
+        Assert::notNull($order, 'Order has not been found.');
+        Assert::notSame(OrderInterface::STATE_CART, $order->getState(), 'Only orders can be updated.');
```
Comment (lchrusciel) [inline src/Handler/Order/UpdatePaymentMethodHandler.php]: Not 100% true, I suppose. Order's in fulfilled state should be blocked as well. 

```suggestion
        Assert::same(OrderInterface::STATE_NEW, $order->getState(), 'Only new orders can be updated.');
```

Comment (mamazu) [reply src/Handler/Order/UpdatePaymentMethodHandler.php]: The validator checks if the payment you are trying to edit was not payed yet. This would be a safer option as an assertion?
